doc: add early warning for policy-default-local-cluster (#40108)
cilium-team/cilium: Documentation/operations/upgrade.rst

* The Helm value of ``enableIPv4Masquerade`` in ``eni`` mode changes from ``true`` to ``false`` by default from 1.18.
  To keep the ``enableIPv4Masquerade`` enabled, explicitly set the value for
  this option to ``true``, or use a value strictly lower than 1.18 for
  ``upgradeCompatibility``.
* This Cilium version now requires a v5.10 Linux kernel or newer.
* CiliumIdentity CRD does not contain Security Labels in metadata anymore except for the namespace label.
* The support for Envoy Go Extensions (proxylib) is deprecated, and will be removed in a future release.
* The kube_proxy_healthz endpoint no longer requires Kubernetes control plane connectivity to succeed.

> * A new opt-out option ``policy-default-local-cluster`` has been added in Cilium 1.18 and will become
>   a default option in Cilium 1.19. If you are using Cilium ClusterMesh and network policies, you will
>   need to take action to update your network policies to avoid this change from breaking connectivity
>   for applications across different clusters. There is no need to do anything for the Cilium 1.17 to 1.18
>   upgrade but we strongly recommend to see :ref:`change_policy_default_local_cluster` for details and
>   migrations recommendations to update your network policies in Cilium 1.18 for the Cilium 1.19 upgrade.

Reviewer: One alternative proposal below. The main idea being to shift the focus from the new flag, which is not that relevant for the users given that it does not modify the current behavior, to the intended behavioral change, to give them more context.

```suggestion
* In a Cluster Mesh environment, network policy ingress and egress selectors currently select by default endpoints from all clusters, unless one or more clusters are explicitly specified in the policy itself. The new ``policy-default-local-cluster`` flag allows to change this behavior, and only select endpoints from the local cluster, unless explicitly specified, to improve the default security posture. This option is intended to become the default in Cilium v1.19. If you are using Cilium ClusterMesh and network policies, you need to take action to update your network policies to avoid this change from breaking connectivity for applications across different clusters. There is no need to do anything for the Cilium 1.17 to 1.18 upgrade, but it is strongly recommend to check :ref:`change_policy_default_local_cluster` for details and migration recommendations to update your network policies in advance for the Cilium 1.19 upgrade.
```

Author: This looks great to me I added it almost verbatim with some very minor fixes

* CiliumIdentity CRD does not contain Security Labels in metadata anymore except for the namespace label.
* The support for Envoy Go Extensions (proxylib) is deprecated, and will be removed in a future release.
* The kube_proxy_healthz endpoint no longer requires Kubernetes control plane connectivity to succeed.

> * A new opt-out option ``policy-default-local-cluster`` has been added in Cilium 1.18 and will become a default option in Cilium 1.19.

Reviewer: I'm somewhat hesitant at already mentioning this in the current form as it may "scare" users, considering that there is no breaking change at the moment, nor the need for any tuning in v1.18. One alternative proposal below (Joe may have a better idea though): 

> In a Cluster Mesh environment, network policy `ingress` and `egress` selectors currently select by default endpoints from all clusters, unless one or more clusters are explicitly specified in the policy itself. The newly introduced `policy-default-local-cluster` flag allows to change this behavior, and only select endpoints from the local cluster, unless explicitly specified, to improve the default security posture. This option is intended to become the default in a future release. [...]

I'm also a bit on the fence on whether to already mention that it will be the default option in 1.19. On the one hand, that's probably the only way to encourage people to take action. On the other, historically we have not been that consistent at following up with this type of actions (for multiple reasons), ending up with e.g., flags deprecated and supposed to be removed in version 1.X being actually dropped in 1.X+5. Maybe a more generic *in a future release* may be a bit safer from that point of view.

Author: > On the other, historically we have not been that consistent at following up with this type of actions (for multiple reasons), ending up with e.g., flags deprecated and supposed to be removed in version 1.X being actually dropped in 1.X+5. Maybe a more generic in a future release may be a bit safer from that point of view.

I mean for us it's super easy timeline wise to not forget, because as soon as we branch out in just a few weeks we can just change the default value of this. Now for actually deprecating/removing the option yes probably we would need to say in a future release :sweat_smile: as it would probably need to be removed in at least 1.20 and probably 1.21 if we want to be generous on migration time.

Reviewer: > I mean for us it's super easy timeline wise to not forget, because as soon as we branch out in just a few weeks we can just change the default value of this.

Yeah, I agree. Just mentioning because although unlikely things may still come up and suggest postponing it (e.g., community feedback, unexpected problems, ...), v1.19 feature freeze is 6 months away after all. I don't have a strong preference though, whatever approach works.

Author: Fair enough! I would prefer calling out that this is a breaking change in 1.19 so that people start to actively consider it ~ now and we could always push back this to later before the 1.19 freeze anyway

Author: I re-read your first comment and just to point out this:
>  nor the need for any tuning in v1.18

Yes users don't need any tuning directly in 1.18 but for most people affected updating policies in 1.18 to prepare for the 1.19 upgrade would be best (or at least consider it, forcing it to false in 1.19 to make this migration over two or more versions in case there is many many policies to update/check works too).

Reviewer: I'm personally still a bit on the fence on how to present this to the users, given that strictly speaking it is not a v1.17 -> v1.18 upgrade note. Anyways, I'd suggest updating the sentence above to give context to the user regarding what this change is about (one example in the initial comment, but feel free to go for something different if you prefer), rather than just mentioning that a new flag got added (especially because the default matches the current behavior). I would also expand a bit the sentence below to clarify that this is not a breaking change in v1.17 -> v1.18, but it will potentially be for v1.18 -> v1.19 (as in, no need to take action before upgrading to v1.18, as the current wording implies).

Author: I tried clarifying that and ok I see your point it's not a problem for me to have this warning elsewhere as long as it's in a sufficiently visible place for people to notice, not sure how this would look like though :sweat_smile:.

>   If you are using Cilium ClusterMesh and network policies enabling this option would certainly result in a breaking change for your existing policies!

Reviewer 2: ```suggestion
  If you are using Cilium ClusterMesh and network policies, you must take action to update your network policies to avoid this change from breaking connectivity for applications across different clusters.
```

>   See :ref:`policies_clustermesh` for details and migrations recommendations.

Reviewer 2: I don't see the migration recommendation, am I missing something?

Author: It's sort of implicitly said there but yes will do a dedicated section like you suggested in your other comment!

   This will make Cilium read and write identities only to CRDs.
#. You can now decommission the KVStore.

.. _change_policy_default_local_cluster:

Preparing for a ``policy-default-local-cluster`` change
#######################################################

Cilium network policies used to implicitly select endpoints from all the clusters.
In Cilium 1.18, we are introducing a new option called ``policy-default-local-cluster``

> which will be set by default in Cilium 1.19. This option restrict endpoints

Reviewer: ```suggestion
which will be set by default in Cilium 1.19. This option restricts endpoints
```

discover all the policies that will change as a result of changing ``policy-default-local-cluster``.
You can also replace ``--all-namespaces`` with ``-n my-namespace`` if you want to only inspect
policies from a particular namespace.

Let's take an imaginary situation where we would have one network policy to update:

.. code-block:: shell-session

    $ cilium clustermesh prepare-policy-default-local-cluster --all-namespaces

    ⚠️ CiliumNetworkPolicy 0/1
            ⚠️ svc1/allow-from-svc2

    ✅ CiliumClusterWideNetworkPolicy 0/0

    ✅ NetworkPolicy 0/0

In this situation you have only one CiliumNetworkPolicy which will be affected by a
``policy-default-local-cluster`` change. Let's take a look at the policy:

.. code-block:: yaml

    apiVersion: "cilium.io/v2"
    kind: CiliumNetworkPolicy
    metadata:
      name: allow-from-svc2
      namespace: svc1
    spec:
      description: "Allow ingress traffic from svc2"
      endpointSelector:
        matchLabels:
          name: svc1
      ingress:
      - fromEndpoints:
        - matchLabels:
            name: svc2

This network policy does not explicitly select a cluster. This means that with ``policy-default-local-cluster``

> sets to ``false`` it will allow traffic coming from svc1 in any clusters connected in your ClusterMesh.

Reviewer: I assume `s/svc1/sv2/`. Same below. I wonder if `foo`, `bar` and alike could be clearer.

   This will make Cilium read and write identities only to CRDs.
#. You can now decommission the KVStore.

.. _change_policy_default_local_cluster:

Preparing for a ``policy-default-local-cluster`` change
#######################################################

Cilium network policies used to implicitly select endpoints from all the clusters.

> In Cilium 1.18, we are introducing a new option called ``policy-default-local-cluster``

Reviewer 2: ```suggestion
Cilium 1.18 introduced a new option called ``policy-default-local-cluster``
```

which will be set by default in Cilium 1.19. This option restricts endpoints
selection to the local cluster by default. If you are using ClusterMesh and
network policies this will be a **breaking change** and you **need to take action**
before upgrading to Cilium 1.19.

> This new option can be set in the ConfigMap or via the Helm value ``clustermesh.policyDefaultLocalCluster``
> and you can of course still force ``policy-default-local-cluster`` to ``false`` in Cilium 1.19 to keep
> the existing behavior. Note that we will deprecate and eventually remove this option in a future release

Reviewer 2: ```suggestion
This new option can be set in the ConfigMap or via the Helm value ``clustermesh.policyDefaultLocalCluster``.
You can of set ``policy-default-local-cluster`` to ``false`` in Cilium 1.19 to keep
the existing behavior, however this option will be deprecated and eventually removed in a future release
```

> so we encourage you to actively plan your migration to set ``policy-default-local-cluster`` to ``true``.

Reviewer 2: ```suggestion
so you should plan your migration to set ``policy-default-local-cluster`` to ``true``.
```

Migrating network policies in practice
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The command ``cilium clustermesh inspect-policy-default-local-cluster --all-namespaces`` can help you
discover all the policies that will change as a result of changing ``policy-default-local-cluster``.
You can also replace ``--all-namespaces`` with ``-n my-namespace`` if you want to only inspect
policies from a particular namespace.

> Let's take an imaginary situation where we would have one network policy to update:

Reviewer 2: ```suggestion
Below is an example where there is one network policy that needs to be updated:
```

.. code-block:: shell-session

    $ cilium clustermesh prepare-policy-default-local-cluster --all-namespaces

    ⚠️ CiliumNetworkPolicy 0/1
            ⚠️ default/allow-from-bar

    ✅ CiliumClusterWideNetworkPolicy 0/0

    ✅ NetworkPolicy 0/0

> In this situation you have only one CiliumNetworkPolicy which will be affected by a

Reviewer: ```suggestion
In this situation you have only one CiliumNetworkPolicy which is affected by a
```

``policy-default-local-cluster`` change. Let's take a look at the policy:

.. code-block:: yaml

    apiVersion: "cilium.io/v2"
    kind: CiliumNetworkPolicy
    metadata:
      name: allow-from-bar
      namespace: default
    spec:
      description: "Allow ingress traffic from bar"
      endpointSelector:
        matchLabels:
          name: foo
      ingress:
      - fromEndpoints:
        - matchLabels:
            name: bar

This network policy does not explicitly select a cluster. This means that with ``policy-default-local-cluster``

> sets to ``false`` it will allow traffic coming from bar in any clusters connected in your ClusterMesh.

Reviewer 2: I wonder whether _foo_ and _bar_ should be somehow emphasized in in the text, to make it clear that they refer to the same foo / bar from the policy. I'm not sure if we have any precedent for this in existing documentation though, so I suppose it's fine to leave it as is.

Author: I made them quoted everywhere now as it seems reasonable to do so but if you have other ideas about that let me know!

> sets to ``false`` it will allow traffic coming from bar in any clusters connected in your ClusterMesh.

Reviewer: ```suggestion
set to ``false`` it allows traffic coming from bar in any clusters connected in your ClusterMesh.
```

> With ``policy-default-local-cluster`` sets to ``true`` (which will be set by default in Cilium 1.19),

Reviewer 2: nit: plural and we probably don't need so many references to v1.19:
```suggestion
With ``policy-default-local-cluster`` set to ``true``,
```

The upgrade notes and the intro to this section already refer to v1.19, so it doesn't need to be repeated here.

> this policy will allow traffic from bar from only the local cluster instead.

Reviewer: ```suggestion
this policy allows traffic from bar from only the local cluster instead.
```

> If foo and bar are always in the same cluster you can set ``policy-default-local-cluster`` to ``true``
> or wait for Cilium 1.19 where it will be set by default.

Reviewer 2: ```suggestion
If foo and bar are always in the same cluster, no further action is necessary.
```

bar is located on a remote cluster you can update your policy like that:

.. code-block:: yaml

    apiVersion: "cilium.io/v2"
    kind: CiliumNetworkPolicy
    metadata:
      name: allow-from-bar
      namespace: default
    spec:
      description: "Allow ingress traffic from bar"
      endpointSelector:
        matchLabels:
          name: foo
      ingress:
      - fromEndpoints:
        - matchLabels:
            name: bar
            io.cilium.k8s.policy.cluster: fixme-cluster-name

If bar is located in multiple cluster you can also use a ``matchExpressions``
selecting multiple clusters like that:

.. code-block:: yaml

    apiVersion: "cilium.io/v2"
    kind: CiliumNetworkPolicy
    metadata:
      name: allow-from-bar
      namespace: default
    spec:
      description: "Allow ingress traffic from bar"
      endpointSelector:
        matchLabels:
          name: foo
      ingress:
      - fromEndpoints:
        - matchLabels:
            name: bar
          matchExpressions:
            - key: io.cilium.k8s.policy.cluster
              operator: In
              values:
                - fixme-cluster-name-1
                - fixme-cluster-name-2

> Alternatively, you can also allow traffic from bar located in every clusters and restore

Reviewer 2: ```suggestion
Alternatively, you can also allow traffic from bar located in every cluster and restore
```
